code/modules/projectiles/guns/projectile/shotgun.dm

```diff
@@ -80,16 +80,94 @@
 	desc = "A sturdy shotgun with a longer magazine and a fixed tactical stock designed for non-lethal riot control."
 	icon_state = "riotshotgun"
 	mag_type = /obj/item/ammo_box/magazine/internal/shot/riot
-	sawn_desc = "Come with me if you want to live."
+	var/style = 1
 
 /obj/item/weapon/gun/projectile/shotgun/riot/attackby(obj/item/A, mob/user, params)
 	..()
-	if(istype(A, /obj/item/weapon/circular_saw) || istype(A, /obj/item/weapon/gun/energy/plasmacutter))
-		sawoff(user)
-	if(istype(A, /obj/item/weapon/melee/energy))
-		var/obj/item/weapon/melee/energy/W = A
-		if(W.active)
-			sawoff(user)
+	if(istype(A, /obj/item/device/riot_upgrade))
+		var/obj/item/device/riot_upgrade/C = A
+		if(C.style == style)
+			to_chat(user, "<span class='notice'>This [src] is already set up in this style!</span>")
+			return
+		if(istype(src.loc, /obj/item/weapon/storage))
+			to_chat(user, "<span class='info'>How do you plan to modify [src] while it's in a bag.</span>")
+			return
+		if(magazine.ammo_count() || chambered)
+			afterattack(user, user)
+			user.visible_message("<span class='danger'>[src] goes off!</span>", "<span class='userdanger'>[src] goes off in your face!</span>")
+			return
+		if(do_after(user, 10, target = src))
+			var/temp = style
+			style = C.style
+			if(style == 1)

```

Comment:
Everything after this point strongly indicates that this code should not be in the attackby of riot shotguns, but instead, on the afterattack of riot upgrades

Reply:
I'm not quite sure what you mean, but if you look at the detective revolver, and the healthanalyzer, both have their upgrade code in the attackby. While they are both relatively simple compared to this code, I do not see why it should be separated or coded so differently.
https://github.com/ParadiseSS13/Paradise/blob/06bb3527c7ec597223235476103452ab518a2e58/code/game/objects/items/devices/scanners.dm#L273
https://github.com/ParadiseSS13/Paradise/blob/c1c4b08807c0eb23e7e2c89ed50310f4c7551f88/code/modules/projectiles/guns/projectile/revolver.dm#L115

Reply:
This isn't the same as either of those cases. You've introduced a specific modkit to the code, and it does literally nothing but serve as an istype() check. Revolvers are modified by touching them with a regular screwdriver, and to be honest, health analyzer code is a piece of shit that should never ever be referenced.

Reply:
If you look at the screwdriver, or even the other tools, they do not have code for construction, everything they do is either in the code of the object like I have here, or relating to them doing damage or color changing on spawn. If you look at stock parts, while I know only partially applicable, all just have one variable, rating, that effects what they do, everything else is done by the machine they are inserted into. Even the sawing this code replaces was in the attackby code of the riot shotgun, quite literally replacing it.

Reply:
Literally all of the examples you have given are different cases

1) Detective revolver has an istype() for screwdrivers
1A) This is okay- The reason detective revolvers are not in the screwdriver code, but rather, screwdrivers are in the revolver code, is because screwdrivers have multiple uses outside of revolvers, and it makes no sense for them to have snowflake code to perform operations on each object they interact with

2) Health scanners have an istype() for health module analyzers
2A) This isn't an excuse because scanner code is awful and this code is wrong, it should be the other way around

3) Stock parts have one variable and machines handle it from there
3A) Stock parts can be used on any machine, and, also, neither the specific machine nor the stock parts actually handle each other, the base machine type has a global implementation for it, which is correct in this case because stock parts have so many subtypes that affect different machines in different ways - which the machines themselves are supposed to define

4) Saw-off code was in attackby
4A) Yes, because circular saws have their own special attack code for other reasons, and their primary usage is not for this object

The consistent thread through here is the usage of the parts being widespread - your specific typepath is not widespread, and it will never be used for anything but this specific weapon for this specific purpose. As such, you should implement the modification behavior on that specific object, instead of making the shotgun behave differently based on the object. 

This won't be merged until you do so.